HTTP_API.md

```diff
@@ -1368,6 +1376,222 @@ The handler-specific descriptions below describe how each handler populates the
+
+    ##### request
+    `GET /api/v2/rules/:name`
+
+    ##### response
+    `200` - The result is a JSON string representing the rule definition.
+
+    `401` - User authentication failed. The reason is an error message.
+    There will be an `X-WWW-Authenticate: $SCHEME` header that indicates
+    the authentication scheme that is used.
+
+    `404` - No rule with the given name exists.
+
+    `500` - There was an unexpected error.
+
+    ##### example
+    ```
+    $ curl http://0.0.0.0:8181/api/v2/rules/Test_Rule
+    {"meta":{"type":"application/json","status":"OK"},"data":{"result":{"name":"Test_Rule","description":"This is a rule for testing","targetAlias":"io.cryostat.Cryostat","eventSpecifier":"template=Continuous,type=TARGET","archivalPeriodSeconds":30,"preservedArchives":1,"maxAgeSeconds":30,"maxSizeBytes":-1}}}
+    ```
+
+* #### `RulesGetHandler`
+
+    ##### synopsis
+    Get a JSON array representing all rule definitions.
+
+    ##### request
+    `GET /api/v2/rules`
+
+    ##### response
+    `200` - The result is a JSON array representing all rule definitions.
+
+    `401` - User authentication failed. The reason is an error message.
+    There will be an `X-WWW-Authenticate: $SCHEME` header that indicates
+    the authentication scheme that is used.
+
+    `500` - There was an unexpected error.
+
+    ##### example
+    ```
+    $ curl http://0.0.0.0:8181/api/v2/rules
+    {"meta":{"type":"application/json","status":"OK"},"data":{"result":[{"name":"Test_Rule","description":"This is a rule for testing","targetAlias":"io.cryostat.Cryostat","eventSpecifier":"template=Continuous,type=TARGET","archivalPeriodSeconds":30,"preservedArchives":1,"maxAgeSeconds":30,"maxSizeBytes":-1}]}}
+    ```
+
+### Stored Target Credentials
+
+* #### `TargetCredentialsPostHandler`
+
+    ##### synopsis
+    Creates stored credentials for a given target. These are used for automated
+    rules processing - if a Target JVM requires JMX authentication, Cryostat
+    will use stored credentials when attempting to open JMX connections to the
+    target. These are retained in Cryostat's memory only and not persisted to
+    disk.
+
+    ##### request
+    `POST /api/v2/targets/:targetId/credentials`
+
+    The request should be an HTTP form with the attributes `"username"` and
+    `"password"`. Both are required.
+
+    ##### response
+    `200` - The result is null. The request was processed successfully and the
+    credentials were stored, potentially overriding previous credentials for the
+    same target.
+
+    `400` - `"username"` and/or `"password"` were not provided.
+
+    `401` - User authentication failed. The reason is an error message.
+    There will be an `X-WWW-Authenticate: $SCHEME` header that indicates
+    the authentication scheme that is used.
+
+    `500` - There was an unexpected error.
+
+    ##### example
+    ```
+    $ curl -F username=user -F password=pass http://0.0.0.0:8181/api/v2/targets/localhost/credentials
```

Comment:
If I try the following with cryostat smoketest, I get a `427`. How do I properly store credentials?
```
$ curl -v --insecure -F username=(omitted) -F password=(omitted) https://0.0.0.0:8181/api/v2/targets/cryostat:9091/credentials 

HTTP/1.1 200 OK

$ curl -v --insecure https://0.0.0.0:8181/api/v2/targets/cryostat:9091/recordingOptionsList

HTTP/1.1 427 Client Error (427)
```

Reply:
You appear to have stored them correctly, but the stored credentials are only (currently) used into automated rules processing. When an external client (like your curl) makes a request, it is still required to provide credentials. I tried to document that in the synopsis for the endpoint:

```
Creates stored credentials for a given target. These are used for automated
rules processing - if a Target JVM requires JMX authentication, Cryostat
will use stored credentials when attempting to open JMX connections to the
target.
```

Reply:
I'm happy with that explanation. If the stored credentials change in the future to support other endpoints we can always update the description.